File: c/transformations_test.c
#include "transformations.h"

#include <assert.h>
#include <math.h>
#include <stdio.h>

#include "matrix.h"
#include "tuple.h"

#define _USE_MATH_DEFINES

// Remove this when I figure out how to link against the library correctly

#define M_PI_2 1.57079632679489661923   // pi/2
#define M_PI_4 0.785398163397448309616  // pi/4
#define M_SQRT2 1.41421356237309504880  // sqrt(2)

void test_translatePoint() {
    matrix t = translation(5, -3, 2);
    tuple p = newPoint(-3, 4, 5);

    tuple r = multiplyMatrixTuple(t, p);

    assert(equalTuple(r, newPoint(2, 1, 7)));
}

void test_translatePoint_reverse() {
    matrix t = translation(5, -3, 2);
    matrix t_inv = inverse(t);
    tuple p = newPoint(-3, 4, 5);

    tuple r = multiplyMatrixTuple(t_inv, p);

    assert(equalTuple(r, newPoint(-8, 7, 3)));
}

void test_translateVector() {
    matrix t = translation(5, -3, 2);
    tuple v = newVector(-3, 4, 5);

    tuple r = multiplyMatrixTuple(t, v);

    assert(equalTuple(r, v));
}

void test_scalePoint() {
    matrix s = scaling(2, 3, 4);
    tuple p = newPoint(-4, 6, 8);

    tuple r = multiplyMatrixTuple(s, p);

    assert(equalTuple(r, newPoint(-8, 18, 32)));
}

void test_scaleVector() {
    matrix s = scaling(2, 3, 4);
    tuple v = newVector(-4, 6, 8);

    tuple r = multiplyMatrixTuple(s, v);

    assert(equalTuple(r, newVector(-8, 18, 32)));
}

void test_scaleVector_reverse() {
    matrix s = scaling(2, 3, 4);
    matrix s_inv = inverse(s);
    tuple v = newVector(-4, 6, 8);

    tuple r = multiplyMatrixTuple(s_inv, v);

    assert(equalTuple(r, newVector(-2, 2, 2)));
}

void test_reflectionIsNegativeScaling() {
    matrix s = scaling(-1, 1, 1);
    tuple p = newPoint(2, 3, 4);

    tuple r = multiplyMatrixTuple(s, p);

    assert(equalTuple(r, newPoint(-2, 3, 4)));
}

void test_rotationX() {
    matrix half_quarter = rotationX(M_PI_4);
    matrix full_quarter = rotationX(M_PI_2);
    tuple p = newPoint(0, 1, 0);

    tuple r1 = multiplyMatrixTuple(half_quarter, p);
    tuple r2 = multiplyMatrixTuple(full_quarter, p);

    assert(equalTuple(r1, newPoint(0, M_SQRT2 / 2, M_SQRT2 / 2)));
    assert(equalTuple(r2, newPoint(0, 0, 1)));
}

void test_rotationX_reverse() {
    matrix half_quarter = rotationX(M_PI_4);
    matrix r_inv = inverse(half_quarter);
    tuple p = newPoint(0, 1, 0);

    tuple r1 = multiplyMatrixTuple(r_inv, p);

    assert(equalTuple(r1, newPoint(0, M_SQRT2 / 2, -1 * M_SQRT2 / 2)));
}

void test_rotationY() {
    matrix half_quarter = rotationY(M_PI_4);
    matrix full_quarter = rotationY(M_PI_2);
    tuple p = newPoint(0, 0, 1);

    tuple r1 = multiplyMatrixTuple(half_quarter, p);
    tuple r2 = multiplyMatrixTuple(full_quarter, p);

    assert(equalTuple(r1, newPoint(M_SQRT2 / 2, 0, M_SQRT2 / 2)));
    assert(equalTuple(r2, newPoint(1, 0, 0)));
}

void test_rotationZ() {
    matrix half_quarter = rotationZ(M_PI_4);
    matrix full_quarter = rotationZ(M_PI_2);
    tuple p = newPoint(0, 1, 0);

    tuple r1 = multiplyMatrixTuple(half_quarter, p);
    tuple r2 = multiplyMatrixTuple(full_quarter, p);

    assert(equalTuple(r1, newPoint(-1 * M_SQRT2 / 2, M_SQRT2 / 2, 0)));
    assert(equalTuple(r2, newPoint(-1, 0, 0)));
}

void test_shearing_xtoy() {
    matrix s = shearing(1, 0, 0, 0, 0, 0);
    tuple p = newPoint(2, 3, 4);

    tuple r = multiplyMatrixTuple(s, p);

    assert(equalTuple(r, newPoint(5, 3, 4)));
}

void test_shearing_xtoz() {
    matrix s = shearing(0, 1, 0, 0, 0, 0);
    tuple p = newPoint(2, 3, 4);

    tuple r = multiplyMatrixTuple(s, p);

    assert(equalTuple(r, newPoint(6, 3, 4)));
}

void test_shearing_ytox() {
    matrix s = shearing(0, 0, 1, 0, 0, 0);
    tuple p = newPoint(2, 3, 4);

    tuple r = multiplyMatrixTuple(s, p);

    assert(equalTuple(r, newPoint(2, 5, 4)));
}

void test_shearing_ytoz() {
    matrix s = shearing(0, 0, 0, 1, 0, 0);
    tuple p = newPoint(2, 3, 4);

    tuple r = multiplyMatrixTuple(s, p);

    assert(equalTuple(r, newPoint(2, 7, 4)));
}

void test_shearing_ztox() {
    matrix s = shearing(0, 0, 0, 0, 1, 0);
    tuple p = newPoint(2, 3, 4);

    tuple r = multiplyMatrixTuple(s, p);

    assert(equalTuple(r, newPoint(2, 3, 6)));
}

void test_shearing_ztoy() {
    matrix s = shearing(0, 0, 0, 0, 0, 1);
    tuple p = newPoint(2, 3, 4);

    tuple r = multiplyMatrixTuple(s, p);

    assert(equalTuple(r, newPoint(2, 3, 7)));
}

void test_individual_transformations() {
    tuple p = newPoint(1, 0, 1);
    matrix A = rotationX(M_PI_2);
    matrix B = scaling(5, 5, 5);
    matrix C = translation(10, 5, 7);

    tuple p2 = multiplyMatrixTuple(A, p);
    tuple p3 = multiplyMatrixTuple(B, p2);
    tuple p4 = multiplyMatrixTuple(C, p3);

    assert(equalTuple(p2, newPoint(1, -1, 0)));
    assert(equalTuple(p3, newPoint(5, -5, 0)));
    assert(equalTuple(p4, newPoint(15, 0, 7)));
}

void test_chained_transformations() {
    tuple p = newPoint(1, 0, 1);
    matrix A = rotationX(M_PI_2);
    matrix B = scaling(5, 5, 5);
    matrix C = translation(10, 5, 7);

    matrix T = multiplyMatrices(C, multiplyMatrices(B, A));
    tuple r = multiplyMatrixTuple(T, p);

    assert(equalTuple(r, newPoint(15, 0, 7)));
}

void test_fluent_transformations() {
    tuple p = newPoint(1, 0, 1);

    matrix T =
        translate(10, 5, 7, scale(5, 5, 5, rotateX(M_PI_2, identityMatrix)));
    tuple r = multiplyMatrixTuple(T, p);

    assert(equalTuple(r, newPoint(15, 0, 7)));
}

int main() {
    test_translatePoint();
    test_translatePoint_reverse();
    test_translateVector();
    test_scalePoint();
    test_scaleVector();
    test_scaleVector_reverse();
    test_reflectionIsNegativeScaling();
    test_rotationX();
    test_rotationX_reverse();
    test_rotationY();
    test_rotationZ();
    test_shearing_xtoy();
    test_shearing_xtoz();
    test_shearing_ytox();
    test_shearing_ytoz();
    test_shearing_ztox();
    test_shearing_ztoy();
    test_individual_transformations();
    test_chained_transformations();
    test_fluent_transformations();
    printf("Transformations module tests successful\n");
}

File: go/matrix/transformations_test.go
package matrix

import (
	"math"
	"raytracerchallenge/tuple"
	"testing"
)

func TestTranslatePoint(t *testing.T) {
	mat := Translation(5, -3, 2)
	p := tuple.NewPoint(-3, 4, 5)

	r := mat.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(2, 1, 7)) {
		t.Fatal("Translation is wrong")
	}
}

func TestTranslatePoint_reverse(t *testing.T) {
	mat := Translation(5, -3, 2)
	tInv, err := mat.Inverse()
	if err != nil {
		t.Fatal("Inverse was wrong")
	}

	p := tuple.NewPoint(-3, 4, 5)

	r := tInv.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(-8, 7, 3)) {
		t.Fatal("Translation is wrong")
	}
}

func TestTranslateVector(t *testing.T) {
	mat := Translation(5, -3, 2)
	v := tuple.NewVector(-3, 4, 5)

	r := mat.MultiplyTuple(v)

	if !r.Equals(v) {
		t.Fatal("Translation is wrong")
	}
}

func TestScalePoint(t *testing.T) {
	s := Scaling(2, 3, 4)
	p := tuple.NewPoint(-4, 6, 8)

	r := s.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(-8, 18, 32)) {
		t.Fatal("Translation is wrong")
	}
}

func TestScaleVector(t *testing.T) {
	s := Scaling(2, 3, 4)
	v := tuple.NewVector(-4, 6, 8)

	r := s.MultiplyTuple(v)

	if !r.Equals(tuple.NewVector(-8, 18, 32)) {
		t.Fatal("Scaling a vector is wrong")
	}
}

func TestScaleVector_reverse(t *testing.T) {
	s := Scaling(2, 3, 4)
	sInv, err := s.Inverse()
	if err != nil {
		t.Fatal("Inverse was wrong")
	}
	v := tuple.NewVector(-4, 6, 8)

	r := sInv.MultiplyTuple(v)

	if !r.Equals(tuple.NewVector(-2, 2, 2)) {
		t.Fatal("Scaling a vector is wrong")
	}
}

func TestReflectionIsNegativeScaling(t *testing.T) {
	s := Scaling(-1, 1, 1)
	p := tuple.NewPoint(2, 3, 4)

	r := s.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(-2, 3, 4)) {
		t.Fatal("Scaling a vector is wrong")
	}
}

func TestRotationX(t *testing.T) {
	halfQuarter := RotationX(math.Pi / 4)
	fullQuarter := RotationX(math.Pi / 2)
	p := tuple.NewPoint(0, 1, 0)

	r1 := halfQuarter.MultiplyTuple(p)
	r2 := fullQuarter.MultiplyTuple(p)

	if !r1.Equals(tuple.NewPoint(0, math.Sqrt2/2, math.Sqrt2/2)) {
		t.Fatal("Half a quarter rotation around X is wrong")
	}
	if !r2.Equals(tuple.NewPoint(0, 0, 1)) {
		t.Fatal("Full quarter rotation around X is wrong")
	}
}

func TestRotationX_reverse(t *testing.T) {
	halfQuarter := RotationX(math.Pi / 4)
	rInv, err := halfQuarter.Inverse()
	if err != nil {
		t.Fatal("Inverse was wrong")
	}

	p := tuple.NewPoint(0, 1, 0)

	r := rInv.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(0, math.Sqrt2/2, -1*math.Sqrt2/2)) {
		t.Fatal("Rotating around X is wrong")
	}
}

func TestRotationY(t *testing.T) {
	halfQuarter := RotationY(math.Pi / 4)
	fullQuarter := RotationY(math.Pi / 2)
	p := tuple.NewPoint(0, 0, 1)

	r1 := halfQuarter.MultiplyTuple(p)
	r2 := fullQuarter.MultiplyTuple(p)

	if !r1.Equals(tuple.NewPoint(math.Sqrt2/2, 0, math.Sqrt2/2)) {
		t.Fatal("Half a quarter rotation around Y is wrong")
	}
	if !r2.Equals(tuple.NewPoint(1, 0, 0)) {
		t.Fatal("Full quarter rotation around Y is wrong")
	}
}

func TestRotationZ(t *testing.T) {
	halfQuarter := RotationZ(math.Pi / 4)
	fullQuarter := RotationZ(math.Pi / 2)
	p := tuple.NewPoint(0, 1, 0)

	r1 := halfQuarter.MultiplyTuple(p)
	r2 := fullQuarter.MultiplyTuple(p)

	if !r1.Equals(tuple.NewPoint(-1*math.Sqrt2/2, math.Sqrt2/2, 0)) {
		t.Fatal("Half a quarter rotation around Z is wrong")
	}
	if !r2.Equals(tuple.NewPoint(-1, 0, 0)) {
		t.Fatal("Full quarter rotation around Z is wrong")
	}
}

func TestShearing_xtoy(t *testing.T) {
	s := Shearing(1, 0, 0, 0, 0, 0)
	p := tuple.NewPoint(2, 3, 4)

	r := s.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(5, 3, 4)) {
		t.Fatal("Shearing X to Y is wrong")
	}
}

func TestShearing_xtoz(t *testing.T) {
	s := Shearing(0, 1, 0, 0, 0, 0)
	p := tuple.NewPoint(2, 3, 4)

	r := s.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(6, 3, 4)) {
		t.Fatal("Shearing X to Z is wrong")
	}
}

func TestShearing_ytox(t *testing.T) {
	s := Shearing(0, 0, 1, 0, 0, 0)
	p := tuple.NewPoint(2, 3, 4)

	r := s.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(2, 5, 4)) {
		t.Fatal("Shearing Y to X is wrong")
	}
}

func TestShearing_ytoz(t *testing.T) {
	s := Shearing(0, 0, 0, 1, 0, 0)
	p := tuple.NewPoint(2, 3, 4)

	r := s.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(2, 7, 4)) {
		t.Fatal("Shearing Y to Z is wrong")
	}
}

func TestShearing_ztox(t *testing.T) {
	s := Shearing(0, 0, 0, 0, 1, 0)
	p := tuple.NewPoint(2, 3, 4)

	r := s.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(2, 3, 6)) {
		t.Fatal("Shearing Z to X is wrong")
	}
}

func TestShearing_ztoy(t *testing.T) {
	s := Shearing(0, 0, 0, 0, 0, 1)
	p := tuple.NewPoint(2, 3, 4)

	r := s.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(2, 3, 7)) {
		t.Fatal("Shearing Z to Y is wrong")
	}
}

func TestIndividualTransformations(t *testing.T) {
	p := tuple.NewPoint(1, 0, 1)
	A := RotationX(math.Pi / 2)
	B := Scaling(5, 5, 5)
	C := Translation(10, 5, 7)

	p2 := A.MultiplyTuple(p)
	p3 := B.MultiplyTuple(p2)
	p4 := C.MultiplyTuple(p3)

	if !p2.Equals(tuple.NewPoint(1, -1, 0)) {
		t.Fatal("Rotation around X is wrong")
	}
	if !p3.Equals(tuple.NewPoint(5, -5, 0)) {
		t.Fatal("Scaling around X is wrong")
	}
	if !p4.Equals(tuple.NewPoint(15, 0, 7)) {
		t.Fatal("Translation around X is wrong")
	}
}

func TestChainedTransformations(t *testing.T) {
	p := tuple.NewPoint(1, 0, 1)
	A := RotationX(math.Pi / 2)
	B := Scaling(5, 5, 5)
	C := Translation(10, 5, 7)

	T := C.Multiply(B).Multiply(A)
	r := T.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(15, 0, 7)) {
		t.Fatal("Chained transformation is wrong")
	}
}

func TestFluentTransformations(t *testing.T) {
	p := tuple.NewPoint(1, 0, 1)

	T := Identity.RotateX(math.Pi/2).Scale(5, 5, 5).Translate(10, 5, 7)
	r := T.MultiplyTuple(p)

	if !r.Equals(tuple.NewPoint(15, 0, 7)) {
		t.Fatal("Fluent transformation is wrong")
	}
}
